### ISSUE TITLE
authenticate_gss_server_step(gss_display_name)

### ISSUE THREAD

**Issue by icrabtree** [dmansfield/passport-negotiate#4]
Hi 
I'm trying to authenticate a user with Kerberos using passport and passport-negotiate on a RHEL IdMClient machine connected to an IdMServer machine. When I first started I was getting the "Unspecified GSS failure" that another person reported (fixed moving the client from RHEL 6.7 to RHEL 7.2). I then moved on to a permission issue on the keytab (fixed with chmod) and then "No key table entry found" by changing from http@idmnuc.nucproxy.com to host@idmnuc.nucproxy.com. Now I’m getting  “Error:  authenticate_gss_server_step(gss_display_name):  A required input parameter could not be read: Unknown error” and I'm not sure what to try next.

Klist –k shows 
Keytab name: FILE:/etc/krb5.keytab
KVNO Principal

---

1 host/idmnuc.nucproxy.com@NUCPROXY.COM
1 host/idmnuc.nucproxy.com@NUCPROXY.COM
1 host/idmnuc.nucproxy.com@NUCPROXY.COM
1 host/idmnuc.nucproxy.com@NUCPROXY.COM

I’m putting the following in for negotiateStrategy
passport.use('kerberos', 
  new negotiateStrategy({enableConstrainedDelegation:false, verbose:true,
    servicePrincipalName: 'host@idmnuc.nucproxy.com'},
      function(principal, done) {
        console.log('negotiating user:' + principal);
        done(null, user)
      })
);

I'm starting the app with
KRB5_KTNAME=/etc/krb5.keytab node app.js

Any idea what could be going on. I’m able to use SSH into a different machine without having to re-login and if I use the RedHat tools for configuring the IdM Server from inside Firefox it works fine using the logged in credentials.

Thanks,
Ian

**Comment by dmansfield** [comment]
The ticket being requested by the client must match the service principal found in the keytab on the server.  It's not enough to change he service principal from 'http' (by the way, should be 'HTTP' not 'http' usually) to 'host' if the client is still obtaining the HTTP/blah.blah.blah ticket. 

One way to check which ticket is being used on the client is by doing a 'fresh' kinit and then hitting the service and immediately doing 'klist' on the client to see which new ticket appeared.

**Comment by icrabtree** [comment]
You are absolutely correct. I figured it out this morning after going through the kerberos package in addition to digging into the MIT Kerberos doc's and the GSS-API docs over the last couple of days. The solution was clear. 

I'm going to give a few more details in case someone stumbles across this issue like I did with the  "Unspecified GSS failure" issue.

What didn't help was that the error I was getting was a total red herring. Turns out that the Kerberos package when calling gss_accept_sec_context() is not checking for GSS_S_CONTINUE_NEEDED since that isn't defined by the GSS-API as an error so GSS_ERROR() doesn't fail. Which caused the code to drop down to gss_display_name() and give the "parameter couldn't be read error". Why the GSS-API is returning GSS_S_CONTINUE_NEEDED instead of an an  error condition to indicate that it couldn't decode the ticket is a mystery. That would have pointed me in the correct direction.

I needed to create a separate keytab for the service HTTP/blah.blah and use that as the servicePrincipalName in the negotiateStrategy() call. Something else that almost screwed me up is that with RHEL IdM it has it's own tools for doing that instead of using kadmin.

I don't know why I didn't notice what klist was doing on the client earlier since I was running it, probably myopia.

At least now I understand a little more about the way Kerberos is working it was just painful getting there.

Thank you for responding and the issue can be closed.
